qubes/tests/integ/vm_qrexec_gui.py

```diff
@@ -65,6 +65,262 @@ def setUp(self):
+        sfreq = 4400
+        audio_in = np.sin(2*np.pi*np.arange(44100)*sfreq/44100)
+        local_user = grp.getgrnam('qubes').gr_mem[0]
+        # Need to use .snd extension so that pw-play (really libsndfile)
+        # recognizes the file as raw audio.
+        if self.testvm1.features['service.pipewire']:
+            record_cmd = ('pw-record --format=f32 --rate=44100 --channels=1 '
+                          'audio_rec.snd')
+            kill_cmd = 'pkill pw-record'
+        else:
+            record_cmd = ('parecord --raw --format=float32le --rate=44100 '
+                          '--channels=1 audio_rec.snd')
+            kill_cmd = 'pkill parecord'
+        record = self.loop.run_until_complete(self.testvm1.run(record_cmd))
+        # give it time to start recording
+        self.loop.run_until_complete(asyncio.sleep(0.5))
+        p = subprocess.Popen(['sudo', '-E', '-u', local_user,
+            'paplay', '--raw', '--format=float32le',
+                    '--rate=44100', '--channels=1'],
+            stdin=subprocess.PIPE)
+        p.communicate(audio_in.astype(np.float32).tobytes())
+        # wait for possible parecord buffering
+        self.loop.run_until_complete(asyncio.sleep(1))
+        self.loop.run_until_complete(self.testvm1.run_for_stdio(kill_cmd))
+        _, record_stderr = self.loop.run_until_complete(record.communicate())
+        if record_stderr:
+            self.fail('parecord printed something on stderr: {}'.format(
+                record_stderr))
+
+        recorded_audio, _ = self.loop.run_until_complete(
+            self.testvm1.run_for_stdio('cat audio_rec.snd'))
+        self.check_audio_sample(recorded_audio, sfreq)
+
+class TC_20_AudioVM_Pulse(TC_00_AudioMixin):
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_220_audio_play_pulseaudio(self):
+        self.prepare_audio_vm('pulseaudio')
+        self.common_audio_playback()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_221_audio_rec_muted_pulseaudio(self):
+        self.prepare_audio_vm('pulseaudio')
+        self.common_audio_record_muted()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_222_audio_rec_unmuted_pulseaudio(self):
+        self.prepare_audio_vm('pulseaudio')
+        self.common_audio_record_unmuted()
+
+    @unittest.skipUnless(spawn.find_executable('parecord'),
+                         "pulseaudio-utils not installed in dom0")
+    def test_223_audio_play_hvm(self):
+        self.testvm1.virt_mode = 'hvm'
+        self.testvm1.features['audio-model'] = 'ich6'
+        self.prepare_audio_vm('pulseaudio')
+        self.loop.run_until_complete(
+            self.testvm1.run_for_stdio('pacmd unload-module module-vchan-sink'))
```

Comment:
This fails with `pipewire-pulseaudio`, which isn't really surprising. I'm okay (at least for now) with skipping this test (and the two below) if no real pulseaudio is installed. The purpose of this test is checking audio passthrough through stubdomain, regardless of the audio stack in the VM. But if there is a simple method to ensure pipewire will use emulated sound card instead of pipewire-qubes module, you can try that too.